framework/src/variables/MooseVariableBase.C

```diff
@@ -171,22 +171,46 @@ std::vector<dof_id_type>
 MooseVariableBase::componentDofIndices(const std::vector<dof_id_type> & dof_indices,
                                        unsigned int component) const
 {
+
+  // Initialize the new_dof_indices vector with the input dof_indices
   std::vector<dof_id_type> new_dof_indices(dof_indices);
+
+  // Only perform mapping if the component is not 0
   if (component != 0)
-  {
-    if (isNodal())
-      for (auto & id : new_dof_indices)
-        id += component;
-    else
-    {
-      unsigned int n = dof_indices.size();
-      for (auto & id : new_dof_indices)
-        id += component * n;
-    }
-  }
+    for (auto & id : new_dof_indices)
+      id += component * _stride[id];
```

Comment:
This looks like the only place we use `_stride`, in this little local loop.  Did I miss someplace elsewhere?

If not, I'd have just used the local APIs to get local information.  That's what `dof_indices` is doing in the first place anyway.  On each DofObject the stride between component 0 and component c for the array variable with base variable number v is `dof_object->n_comp(sys_num,v)*c`.  No global O(N_nodes) vectors or parallel communication should be necessary here.

Reply:
So is it assumed that `componentDofIndices` will only be called one element/dof object at a time?
For example, if someone has the dof index i, and just needs to get the dof index j which is the component c of that array variable would your method work?  I was having trouble because I was trying to make this general enough that given a single  dof or any vector of dofs in any order of the component 0 of the array variable, `componentDofIndices` would return the correct component c dofs that correspond.

Reply:
Yeah, that's a good point.  I looked into the use cases in the framework and it's always in the context of a single `DenseMatrix` jacobian block, but in the context of DG and FV schemes that doesn't have to mean we're on a single element, does it?

I'm going to have to harass @lindsayad to double-check.

Reply:
Methods like `Assembly::addJacobianBlock` which calls `componentDofIndices` are called from "DG" methods like `Assembly::addJacobianNeighbor` like so:
```c++
        addJacobianBlock(_sys.getMatrix(tag),
                         jacobianBlockNeighbor(Moose::ElementNeighbor, i, j, LocalDataKey{}, tag),
                         *ivar,
                         *jvar,
                         ivar->dofIndices(),
                         jvar->dofIndicesNeighbor());

        addJacobianBlock(_sys.getMatrix(tag),
                         jacobianBlockNeighbor(Moose::NeighborElement, i, j, LocalDataKey{}, tag),
                         *ivar,
                         *jvar,
                         ivar->dofIndicesNeighbor(),
                         jvar->dofIndices());

	addJacobianBlock(_sys.getMatrix(tag),
                         jacobianBlockNeighbor(Moose::NeighborNeighbor, i, j, LocalDataKey{}, tag),
                         *ivar,
                         *jvar,
                         ivar->dofIndicesNeighbor(),
                         jvar->dofIndicesNeighbor());
```
so there is cross-element dof coupling

Reply:
If need be we can just use the stride method for anything that is not already correct? Since the old method works perfect for Lagrange and monomial families I keep that part of the code. Any family which that isn't true, use strides?

Reply:
I will have to look at this in detail to answer you questions. If @roystgnr beats me to it, then more power to him

Reply:
One more idea, what if we just change how libmesh adds the dofs in for Hiecharic + other families?
We make sure that the dofs added follow the logic expected before for `isNodal()` It would just change the order we put the dofs into the midside nodes.

Reply:
There is no way to change how libMesh adds DoFs for multiple variables in a group without breaking existing optimizations, optimizations that are saving us factors of n_components for array variables in particular.  I'm not even sure there's a way to change things *with* breaking existing optimizations - the logic for isNodal() is basically assuming that we have one DoF per node, but as soon as you e.g. try to put 16 cubic DoFs onto a Quad9, 16>9 really bites you there.

We could add array variables to libMesh, instead of the hack here to emulate them via multiple variables in a group, but that wouldn't be a quick fix and I'm not sure how many APIs would need changing.

It looks like, although we're not always dealing with DoF indices that are local to a single element at a time, we're always dealing with index vectors that are local to a single element or one of its neighbors?  That would still be just fine for figuring out strides on a local level, if we could pass along which element was associated with each indices vector.

Reply:
the largest "stencil" we support for finite elements is mortar, which has a higher-d "element", higher-d "neighbor", and lower-d "side" whose interior parent is the "element". So this suggestion

> That would still be just fine for figuring out strides on a local level, if we could pass along which element was associated with each indices vector.

seems good to me